### PR TITLE
[chore][receiver/splunkenterprise]Splunkent scrape refactor

### PR DESCRIPTION
**Pull request by shalper2**
**Description:** Refactored the way scrapes work to better utilize golang's language features to parallelize the scraping process.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** Existing tests cover the scrape process. Inclusion in a collector distribution was succesful.

**Documentation:** No changes to documentation

### CODE REVIEW

**Review by greatestusername-splunk — APPROVED** [review]

**Review by greatestusername — APPROVED** [review]

**Comment by shalper2** [comment]
@open-telemetry/collector-contrib-approvers howdy! hoping someone could take a look at this for me

File: receiver/splunkenterprisereceiver/scraper.go [hunk 1]
```diff
@@ -52,36 +53,86 @@ func (s *splunkScraper) start(_ context.Context, h component.Host) (err error) {
...
+			eOut <- errs
+			return
+		case err, ok := <-eQueue:
+			// shutdown
+			if err == nil || !ok {
+				eOut <- errs
+				return
+				// or add an error to errs
+			} else {
+				errs.Add(err)
```
Comment (atoulme) [inline receiver/splunkenterprisereceiver/scraper.go:75]: is that threadsafe?

File: receiver/splunkenterprisereceiver/scraper.go [hunk 2]
```diff
@@ -52,36 +53,86 @@ func (s *splunkScraper) start(_ context.Context, h component.Host) (err error) {
...
+		// context timeout
+		case <-ctx.Done():
+			eOut <- errs
+			return
+		case err, ok := <-eQueue:
+			// shutdown
+			if err == nil || !ok {
+				eOut <- errs
+				return
+				// or add an error to errs
```
Comment (atoulme) [inline receiver/splunkenterprisereceiver/scraper.go:73]:
```suggestion
			// or add an error to errs
```

File: receiver/splunkenterprisereceiver/scraper.go (hunk 1 above)
Comment (atoulme) [reply receiver/splunkenterprisereceiver/scraper.go:75]:
```
// Add adds a regular error.
func (s *ScrapeErrors) Add(err error) {
	s.errs = append(s.errs, err)
}
```

It is not threadsafe.

Comment (atoulme) [reply receiver/splunkenterprisereceiver/scraper.go:75]: How about accumulating errors in the channel and emptying the channel all at once when the scrape is done?

Comment (shalper2) [reply receiver/splunkenterprisereceiver/scraper.go:75]: sorry if im being a little dense but do i need to be worried about the threadsafety of `Add()` in this case? there is only one function which accesses the `errs` object and it does so in a single routine.

File: receiver/splunkenterprisereceiver/scraper.go [hunk 3]
```diff
@@ -52,36 +53,86 @@ func (s *splunkScraper) start(_ context.Context, h component.Host) (err error) {
 	return nil
 }
 
+// listens to the error channel and combines errors sent from different metric scrape functions,
+// returning the combined error list should context timeout or a nil error value is sent in the
+// channel signifying the end of a scrape cycle
+func errorListener(ctx context.Context, eQueue <-chan error, eOut chan<- *scrapererror.ScrapeErrors) {
+	errs := &scrapererror.ScrapeErrors{}
+
+	for {
```
Comment (dmitryax) [inline receiver/splunkenterprisereceiver/scraper.go:62]: Same comment here https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31260#discussion_r1547038751

File: receiver/splunkenterprisereceiver/scraper.go [hunk 4]
```diff
@@ -52,36 +53,86 @@ func (s *splunkScraper) start(_ context.Context, h component.Host) (err error) {
...
+		select {
+		// context timeout
+		case <-ctx.Done():
+			eOut <- errs
+			return
+		case err, ok := <-eQueue:
+			// shutdown
+			if err == nil || !ok {
```
Comment (dmitryax) [inline receiver/splunkenterprisereceiver/scraper.go:70]: Same comment as in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31260#discussion_r1547136649

**Review by dmitryax — APPROVED** [review]

**Pull request merged**